File: VB/DXPivotGrid_CustomSummary/DistinctCountFunction.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.Data.Filtering
Imports DevExpress.PivotGrid.ServerMode.Queryable
Imports System.Linq.Expressions
Imports DevExpress.Data.Linq
Imports DevExpress.DataProcessing.Criteria

Namespace DXPivotGrid_CustomSummary
	Friend Class DistinctCountFunction
		Implements ICustomAggregateFunction, ICustomFunctionOperatorBrowsable

		Public ReadOnly Property Name() As String Implements DevExpress.Data.Filtering.ICustomFunctionOperator.Name
			Get
				Return "DistinctCount"
			End Get
		End Property
		Public ReadOnly Property MinOperandCount() As Integer Implements ICustomFunctionOperatorBrowsable.MinOperandCount
			Get
				Return 1
			End Get
		End Property

		Public ReadOnly Property MaxOperandCount() As Integer Implements ICustomFunctionOperatorBrowsable.MaxOperandCount
			Get
				Return 1
			End Get
		End Property

		Public ReadOnly Property Description() As String Implements ICustomFunctionOperatorBrowsable.Description
			Get
				Return "Displays the number of unique values of the field"
			End Get
		End Property

		Public ReadOnly Property Category() As FunctionCategory Implements ICustomFunctionOperatorBrowsable.Category
			Get
				Return DevExpress.Data.Filtering.FunctionCategory.Math
			End Get
		End Property

		Public ReadOnly Property FunctionCategory() As String
			Get
				Return "Aggregate"
			End Get
		End Property

		Public Function Evaluate(ParamArray ByVal operands() As Object) As Object Implements DevExpress.Data.Filtering.ICustomFunctionOperator.Evaluate
			Throw New NotImplementedException()
		End Function

		Public Function GetAggregationContextType(ByVal inputType As Type) As Type Implements ICustomAggregateFunction.GetAggregationContextType
			Return GetType(DistinctCountAggregateState)
		End Function

		Public Function IsValidOperandCount(ByVal count As Integer) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandCount
			Return count <= MaxOperandCount AndAlso count >= MinOperandCount
		End Function

		Public Function IsValidOperandType(ByVal operandIndex As Integer, ByVal operandCount As Integer, ByVal type As Type) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandType
			Return IsValidOperandCount(operandCount) AndAlso operandIndex = 0
		End Function

		Public Function ResultType(ParamArray ByVal operands() As Type) As Type Implements DevExpress.Data.Filtering.ICustomFunctionOperator.ResultType
			Return GetType(Integer)
		End Function
	End Class
	Friend Class DistinctCountAggregateState
		Implements ICustomAggregateFunctionContext(Of Object, Integer)

		Private values As New HashSet(Of Object)()
		Public Function GetResult() As Integer Implements ICustomAggregateFunctionContext(Of Object, Integer).GetResult
			Return values.Count
		End Function
		Public Sub Process(ByVal value As Object) Implements ICustomAggregateFunctionContext(Of Object, Integer).Process
			If Not values.Contains(value) Then
				values.Add(value)
			End If
		End Sub
	End Class
End Namespace

File: CS/DXPivotGrid_CustomSummary/DistinctCountFunction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.Data.Filtering;
using DevExpress.PivotGrid.ServerMode.Queryable;
using System.Linq.Expressions;
using DevExpress.Data.Linq;
using DevExpress.DataProcessing.Criteria;

namespace DXPivotGrid_CustomSummary {
    class DistinctCountFunction : ICustomAggregateFunction, ICustomFunctionOperatorBrowsable {
        public string Name => "DistinctCount";
        public int MinOperandCount => 1;

        public int MaxOperandCount => 1;

        public string Description => @"Displays the number of unique values of the field";

        public FunctionCategory Category => DevExpress.Data.Filtering.FunctionCategory.Math;

        public string FunctionCategory => "Aggregate";

        public object Evaluate(params object[] operands) {
            throw new NotImplementedException();
        }

        public Type GetAggregationContextType(Type inputType) {
            return typeof(DistinctCountAggregateState);
        }

        public bool IsValidOperandCount(int count) {
            return count <= MaxOperandCount && count >= MinOperandCount;
        }

        public bool IsValidOperandType(int operandIndex, int operandCount, Type type) {
            return IsValidOperandCount(operandCount) && operandIndex == 0;
        }

        public Type ResultType(params Type[] operands) {
            return typeof(int);
        }
    }
    class DistinctCountAggregateState : ICustomAggregateFunctionContext<object, int> {
        HashSet<object> values = new HashSet<object>();
        public int GetResult() {
            return values.Count;
        }
        public void Process(object value) {
            if(!values.Contains(value))
                values.Add(value);
        }
    }
}
